### ISSUE TITLE
Global and local frames

### ISSUE THREAD

**Issue by alessandroscoppio** [ArashJavan/DeepLIO#3]
Hi there!

First off, thanks for sharing such beautiful project! For my thesis I am working on basically the same task, but using more sensors (include RGB images as well) and different processing (especially for the LiDAR, that I am processing as a 3D sparse tensor instead of projecting it to 2D). I liked the waty you structured the train logic a lot!

What I would like to ask you tho, is the reason behind having both a global and local reference when calculating the loss [(here)](https://github.com/ArashJavan/DeepLIO/blob/master/deeplio/models/trainer.py#L210).

In my project I am planning to use only the local ones (e.g. at each step regress the motion from beginning to stop of the step), I tried using global one only but the results are pretty bad.

Hope to have some nice insight!
Cheers

**Comment by alessandroscoppio** [comment]
Hi @ArashJavan , thanks for your reply.

> Somewhere around -20..-30 is also what I get.
Did you mean the total loss? Because that is also what I get, but it feels "wrong" to me. Neural nets training is formulated as a minimization problem and that's why the standard loss functions used are always positive (MSE, cross-entropy etc.). 
Therefore I am worried to see a loss that gets negative, if we were to use just a simple MSE and it would get negative that would mean that we are updating weights such that the difference to groundtruth gets even more.

I am thinking of squaring the whole HWS loss and see how it goes, but I will have a second look at the original paper first, maybe I am missing something..

**Comment by alessandroscoppio** [comment]
Dear @ArashJavan , while renewing my question on how your final loss is looking like, I would like to ask another question (in the same issue to not pollute your github, but it can be moved if you want to keep things separated):
When regressing the orientation, I see that the orientation fully connected layer predicts 3 values and not 4, even though the orientation is expressed as a quaternion. Does this mean that you are only regressing the wx, wy, wz component and calculating the real part (wq)? Because so far I am regressing the full quaternion, and results are actually pretty ok (while the position is still very off), but I can see that this could lead to quaternion that are not unit quaternions.
What's your take on this?

**Comment by ArashJavan** [comment]
@alessandroscoppio , well the relative orientation between two consecutive is predicted not as a quaternion, but as "se(3)", also the Lie-Algebra or log-space of quatenrnion. This also why the f2f-ground-truth is also first converted to the log-space. See [here](https://github.com/ArashJavan/DeepLIO/blob/master/deeplio/models/misc.py#L104).
What I have experienced is that optimizing directly on a normalized quaternion is harder than the log-space, since there are no constraints (normalization).
